modules/miner/update.go

```diff
@@ -1,11 +1,205 @@
 package miner
 
 import (
-	"github.com/NebulousLabs/Sia/encoding"
 	"github.com/NebulousLabs/Sia/modules"
 	"github.com/NebulousLabs/Sia/types"
 )
 
+func (m *Miner) getNewSplitSets(diff *modules.TransactionPoolDiff) []*mapElement {
+	// Split the new sets and add the splits to the list of transactions we pull
+	// form.
+	newElements := make([]*mapElement, 0)
+	for _, newSet := range diff.AppliedTransactions {
+		// Split the sets into smaller sets, and add them to the list of
+		// transactions the miner can draw from.
+		//
+		// TODO: Split the one set into a bunch of smaller sets using the cp4p
+		// splitter.
+		m.setCounter++
+		m.fullSets[newSet.ID] = []int{m.setCounter}
+		var size uint64
+		var totalFees types.Currency
+		for i := range newSet.IDs {
+			size += newSet.Sizes[i]
+			for _, fee := range newSet.Transactions[i].MinerFees {
+				totalFees = totalFees.Add(fee)
+			}
+		}
+		// We will check to see if this splitSet belongs in the block.
+		s := &splitSet{
+			size:         size,
+			averageFee:   totalFees.Div64(size),
+			transactions: newSet.Transactions,
+		}
+
+		elem := &mapElement{
+			set:   s,
+			id:    splitSetID(m.setCounter),
+			index: 0,
+		}
+
+		m.splitSets[splitSetID(m.setCounter)] = s
+		newElements = append(newElements, elem)
+	}
+	return newElements
+}
+
+// addNewTxns adds new unconfirmed transactions to the miner's transaction
+// selection.
+func (m *Miner) addNewTxns(diff *modules.TransactionPoolDiff) {
+	// Get new splitSets (in form of mapElement)
+	newElements := m.getNewSplitSets(diff)
+
+	// Place each elem in one of the MapHeaps
+	for _, elem := range newElements {
+		m.addMapElementTxns(elem)
+	}
+}
+
+// addMapElementTxns places the splitSet from a mapElement into the correct
+// mapHeap.
+func (m *Miner) addMapElementTxns(elem *mapElement) {
+	candidateSet := elem.set
+
+	// Check if heap for highest fee transactions has space.
+	if m.blockMapHeap.size+candidateSet.size < types.BlockSizeLimit-5e3 {
+		m.blockMapHeap.Push(elem)
+		return
+	}
+
+	// The block heap doesn't have enough space for this transaction.
+	// Check if removing sets from the blockMapHeap will be worth it.
+	// bottomSets will hold  the lowest fee sets from the blockMapHeap
+	bottomSets := make([]*mapElement, 0)
+	var sizeOfBottomSets uint64
+	var averageFeeOfBottomSets types.Currency
+
+	// While the heap cannot fit this set s, and while the (weighted)
+	// average fee for the lowest sets from the block is less than
+	// the fee for the set s, continue removing from the heap.
+	for {
```

Comment:
general design here - it's not immediately obvious that this loop is going to exit. It's also not clear that the variables declared above the loop are part of the loop state, because they are separated by a newline. Should take the form:

```
// comment
vars := myVars
for {
```

In loop design, you want to try to make it obvious that the loop is going to exit.